Add support to run stockpile along with collect sysinfo

chaitanyaenr opened the pull request:
Stockpile is as part of collect sysinfo to gather system information
about controller and registered remotes. Eventually, collect sysinfo
will be replaced with stockpile once it has support to gather all
the info that it's collecting.

Fixes #915

chaitanyaenr commented on agent/util-scripts/pbench-collect-sysinfo:
```diff
@@ -242,6 +242,55 @@ if [ $lcl_cnt -gt 0 ]; then
+debug_log "[$script_name]Collecting stockpile data"
+ansible-playbook -vv --extra-vars "stockpile_user=$stockpile_user local_remote_user=$stockpile_local_remote_user host_remote_user=$stockpile_host_remote_user stockpile_output_path=$stockpile_output_path" -i $stockpile_inventory $stockpile_path/stockpile/stockpile.yml 2>&1 > "$stockpile_log"
+if [[ $? != 0 ]]; then
+	error_log "[$script_name] Stockpile run failed"
```
The benchmark run will not fail even if stockpile fails for now.

chaitanyaenr commented on agent/util-scripts/pbench-collect-sysinfo:
```diff
@@ -242,6 +242,55 @@ if [ $lcl_cnt -gt 0 ]; then
+if [[ -d "$stockpile_path/stockpile" ]]; then
+	rm -rf $stockpile_path/stockpile
+fi
+git clone https://github.com/redhat-performance/stockpile.git $stockpile_path/stockpile &>/dev/null
```
Stockpile git repo is cloned for now to get the stockpile code. It would be great if we make a release in stockpile and package it as an rpm. The only problem with this approach is that the stockpile is very fast moving, so we won't get the latest changes.

chaitanyaenr commented on agent/util-scripts/pbench-collect-sysinfo:
```diff
@@ -242,6 +242,55 @@ if [ $lcl_cnt -gt 0 ]; then
+stockpile_output_path=$(getconf.py -C $configfile stockpile_output_path stockpile)
+
+# Install stockpile and its dependencies
+which git &>/dev/null
```
As discussed with @portante and @ndokos, the dependencies should be just installed on the controller, not on all the remotes. We are working on a common approach to handle the dependencies during the runtime.

atheurer commented on agent/base:
```diff
@@ -230,6 +230,17 @@ function generate_inventory {
 }
 
+# generate inventory file with controller and remotes for stockpile
+export stockpile_inventory=/tmp/stockpile_inventory.$$
```
Shouldn't this belong in $pbench_run/tmp?

atheurer commented on agent/util-scripts/pbench-collect-sysinfo:
```diff
@@ -242,6 +242,52 @@ if [ $lcl_cnt -gt 0 ]; then
+		yum install $package -y &>/dev/null
+	fi
+done
+if [[ -d "$stockpile_path/stockpile" ]]; then
```
So we now have to git clone every time we need to run stockpile?
And we're putting the repo under /opt/pbench-agent?  I am not sure /opt/pbench-agent is the best place for ephemeral storage.  It would probably be more appropriate in /var/lib/pbench-agent.

atheurer commented:
@chaitanyaenr can you add the ability to have collect-sysinfo only collect stockpile, in the same way you can specify other collectors, like --sysinfo=stockpile?

atheurer commented:
@chaitanyaenr FYI, this integration IMO does not work that great when just running 'pbench-collect-sysinfo' on the cmdline.  This script requires a --dir to put have a place to put data into, but the playbook for stockpile does not use this.  So, if you run this script, even with --dir, the stockpile json file ends up in "/".  The Ansible playbook should be using $dir, and cannot assume $benchmark_run_dir is available.

chaitanyaenr commented:
@atheurer thanks for the review, added stockpile as an option to collect-sysinfo and also modified it to use $dir instead of $benchmark_run_dir. Running pbench-collect-sysinfo on the cmdline should work now.

chaitanyaenr commented:
@atheurer any more changes needed?

portante commented on agent/base:
```diff
@@ -230,6 +230,17 @@ function generate_inventory {
 	ls $tools_group_dir | awk -F@ '/^remote/ {print $2};'
 }
 
+# generate inventory file with controller and remotes for stockpile
```
This looks like a repeat of an existing function.  Any way we can parameterize this and collapse into one?

chaitanyaenr replied:
Fixed.

portante commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -123,6 +123,58 @@ function collect_ara_data {
+	fi
+
+	# set stockpile log file and output paths
+	if [[ -z "$stockpile_log" ]]; then
```
Where would this reference to `stockpile_log` be satisfied if not from the config file?

portante replied:
@chaitanyaenr, does this question make sense to you?

chaitanyaenr replied:
Yes, modified it to place the log file in sysinfo dir along with the other sysinfo log files.

portante commented on agent/config/pbench-agent.cfg.example:
```diff
@@ -40,8 +40,16 @@ interval = 30
+local_remote_user = root 
+host_remote_user = root
+stockpile_path = /var/lib/pbench-agent
+stockpile_log =
```
Can you add a comment that `stockpile_log` and `stockpile_output_path` are optional, the default being ... (whatever the default is)?

chaitanyaenr replied:
Done.

portante commented on agent/config/pbench-agent.cfg.example:
```diff
@@ -40,8 +40,16 @@ interval = 30
+stockpile_user = root 
+local_remote_user = root 
+host_remote_user = root
+stockpile_path = /var/lib/pbench-agent
```
It looks as if `stockpile_path` is optional in the code, with a hard-coded default of `/var/lib/pbench-agent`.  Can you confirm, and if so, make this a required parameter with a comment that describes why `/var/lib/pbench-agent` is used, and remove the hard-coded default in the code?

chaitanyaenr replied:
You are right @portante , it's optional. Removed the hard-coded default, we use $pbench_run defined in the config if it's not set.

portante commented on agent/config/pbench-agent.cfg.example:
```diff
@@ -40,8 +40,16 @@ interval = 30
 # pandas-package = python-pandas
 # Fedora has python2-pandas and python3-pandas
 # pandas-package = python2-pandas
+stockpile_dependencies = git, ansible
```
Why does stockpile depend on `git`?

chaitanyaenr replied:
We don't have stockpile packaged, so we are cloning the repo for now. Eventually we will be packaging and installing a stable version of stockpile.

portante replied:
We have `check_install_rpm` in the `agent/base` include which should take care of such checks.

But let's change this to assume the stockpile code is "part" of the `pbench-agent` install using `git submodules` [(1)](https://git-scm.com/docs/git-submodule).

Further, it seems best to make the `pbench-agent` RPM depend on ansible now, as we are using it in other places in the code, too.

chaitanyaenr replied:
Agree with making the pbench-agent RPM depend on ansible. Will take a look at git submodules  @portante.

portante replied:
So now that we are using git sub modules, do we still need this?

portante commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -123,6 +123,58 @@ function collect_ara_data {
+        	if [[ $? != 0 ]]; then
+                	debug_log "[$script_name] $package is not installed on the system, installing"
+                	echo "$package is not installed on the system, installing"
+                	yum install $package -y &>/dev/null
```
For now, stockpile is just an ansible playbook.  Are there other functions of pbench-agent that rely on ansible?  If so, can we avoid installing it here, and just reporting to the user that without ansible installed these functions are not available?

portante commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -123,6 +123,58 @@ function collect_ara_data {
+        	fi
+	done
+	if [[ -z "$stockpile_path" ]]; then
+        	stockpile_path=/var/lib/pbench-agent
```
This seems bad to hard-code a default.

chaitanyaenr replied:
Fixed

portante commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -123,6 +123,58 @@ function collect_ara_data {
+	if [[ -d "$stockpile_path/stockpile" ]]; then
+        	rm -rf $stockpile_path/stockpile
+	fi
+	git clone https://github.com/redhat-performance/stockpile.git $stockpile_path/stockpile &>/dev/null
```
Can we use `curl` instead to fetch a release tarball from github?

chaitanyaenr replied:
We don't have a release for stockpile as of now, it would be nice to have a release and maybe an rpm for a stable release like what we have in pbench.

mrsiano replied:
you can add TODO here to handle this, once it's ready.
i'll also open issue for that.

portante replied:
Seems best to work with `git submodules` [1](https://git-scm.com/docs/git-submodule).  Instead of cloning here, when we package the `pbench-agent` we should pull a specific version of `stockpile` into the source tree so that it ends up installed in `/opt/pbench-agent/...` somewhere.

chaitanyaenr commented:
@portante made some changes to address the comments, PTAL.

mrsiano commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -123,6 +123,58 @@ function collect_ara_data {
 
+function collect_stockpile_data {
+	# parse config
+	stockpile_path=$(getconf.py stockpile_path stockpile)
```
is this getconf.py part of pbench utils ?!
also looks like we call it many times, we can do something like single call with arguments and return some sort of json string with with answer per argument ?

mrsiano commented:
this work out of the box @chaitanyaenr can you apply the changes please to merge this one.
/ltgm

portante commented:
@chaitanyaenr, would it be okay if @mrsiano made a PR to address the comments requesting changes, but post that PR against your repo so that you could then merge them, and update this PR with the fixes?

We'd like to get this functionality added sooner than later.

chaitanyaenr commented:
@portante  sure. Most of the comments have been addressed in the latest commit, we can merge @mrsiano's PR with the fixes and get this in, sorry for the delay.

portante commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -139,6 +191,8 @@ for item in ${sysinfo//,/ };do
 	elif [[ "$item" == "ara" ]]; then
 		collect_ara_data &> $pbench_log
+	elif [[ "$item" == "stockpile" ]]; then
+		collect_stockpile_data
```
Shouldn't this invocation be put in the background like the other sysinfo options?

portante commented:
@chaitanyaenr, can you rebase this work and comment on some of the review comments still opened and not resolved?

portante commented on agent/base:
```diff
@@ -244,10 +244,19 @@ function verify_common_bench_script_options {
 export INVENTORY=/tmp/inventory.$$
 trap "rm -f $INVENTORY" QUIT INT EXIT
 function generate_inventory {
```
Seems like we should promote `generate_inventory` to a separate script that lives in `util-scripts`, perhaps ... or ... leverage the new Perl Ansible module that @atheurer recently did instead.

Having the trap in the base is not great because it fires whether or not a caller invoked generate_inventory or not.

portante commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -123,6 +123,50 @@ function collect_ara_data {
+	stockpile_log=$(getconf.py stockpile_log stockpile)
+
+	# install stockpile and its dependencies
+	for package in $stockpile_dependencies; do
```
Seems like this can be removed now that we'll be packaging it all together.

portante commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -123,6 +123,50 @@ function collect_ara_data {
+	# generate inventory and run stockpile
+	generate_inventory stockpile > $INVENTORY
+	debug_log "[$script_name]Collecting stockpile data"
+	ansible-playbook -vv --extra-vars "stockpile_user=$stockpile_user local_remote_user=$stockpile_local_remote_user host_remote_user=$stockpile_host_remote_user stockpile_output_path=$stockpile_output_path" -i $INVENTORY $stockpile_path/stockpile/stockpile.yml 2>&1 > "$stockpile_log"
```
Can you add a `debug_log` line to capture the `ansible-playbook` command line invocation?  Perhaps write the full command to a file, and then invoke the file?  Like we do in benchmark scripts?

chaitanyaenr commented:
Sorry for the delay @portante, got buried in ocp 4.0 scale testing, will take a look at the comments and make changes accordingly.

chaitanyaenr commented:
/hold

portante commented:
> /hold

What does that mean?

chaitanyaenr commented:
@portante missed @aakarshg changes when I rebased. It should be good to go now.

portante commented:
Still waiting for Nick to review before we merge.

ndokos commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -127,6 +127,45 @@ function collect_ara_data {
+	stockpile_log=$(getconf.py stockpile_log stockpile)
+
+	if [[ -z "$stockpile_path" ]]; then
+		stockpile_path=$(getconf.py pbench_install_dir DEFAULT)
```
This does not work: it needs to be `getconf.py pbench_install_dir pbench-agent` or `getconf.py pbench_install_dir stockpile`.

ndokos commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -127,6 +127,45 @@ function collect_ara_data {
+		stockpile_path=$(getconf.py pbench_install_dir DEFAULT)
+	fi
+	if [[ -d "$stockpile_path/stockpile" ]]; then
+        	rm -rf $stockpile_path/stockpile
```
Why is this removed? The playbook is removed along with the directory and the collection fails.

portante replied:
Looks like this was resolved.

ndokos commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -127,6 +127,45 @@ function collect_ara_data {
+	debug_log "[$script_name]Collecting stockpile data"
+	stockpile_opts="stockpile_user=$stockpile_user local_remote_user=$stockpile_local_remote_user host_remote_user=$stockpile_host_remote_user stockpile_output_path=$stockpile_output_path"
+	stockpile_playbook="$stockpile_path/stockpile/stockpile.yml"
+	echo "ansible-playbook -vv --extra-vars $stockpile_opts -i $INVENTORY $stockpile_playbook" >$stockpile_path/stockpile.cmd
```
I think the quoting here is wonky: the `--extra-vars` argument is an unquoted space-separated list of `var=value` pairs. This shoulde be 

    echo "ansible-playbook -vv --extra-vars '"$stockpile_opts"' -i $INVENTORY $stockpile_playbook" >$stockpile_path/stockpile.cmd

Also, I don't think the  command should be in $stockpile_path, which if otherwise unspecified is `/opt/pbench-agent`: that should be considered read-only.

portante replied:
Seems to have been resolved.

portante commented:
@chaitanyaenr, have you had a chance to consider @ndokos's comments?

chaitanyaenr commented:
@portante @ndokos made the changes requested, PTAL. Unit tests related to 'pbench-copy-results' failed, will check. Thanks.

ndokos commented on agent/config/pbench-agent.cfg.example:
```diff
@@ -7,4 +7,4 @@ pbench_web_server = pbench.example.com
 [config]
 path = %(pbench_install_dir)s/config
-files = pbench-agent-default.cfg
+files = pbench-agent.cfg
```
This should be pbench-agent-default.cfg, as it was before.

portante replied:
Looks like this change was taken care of with the latest commit.

portante commented on agent/util-scripts/gold/pbench-collect-sysinfo/test-24.txt:
```diff
@@ -1,5 +1,6 @@
+<<<<<<< HEAD
```
Seems to have been addressed.

portante commented on agent/util-scripts/pbench-sysinfo-dump:
```diff
@@ -134,6 +134,42 @@ function collect_ara_data {
+	stockpile_log=$(getconf.py stockpile_log stockpile)
+
+	if [[ -z "$stockpile_path" ]]; then
+		stockpile_path=$(getconf.py pbench_install_dir pbench-agent)
```
Rather than wait for an update, I'm gonna merge this and post a PR to fix this last comment.

portante replied:
Resolved by PR #1209.

portante commented:
@ndokos, can you sign off on this before we merge it?
